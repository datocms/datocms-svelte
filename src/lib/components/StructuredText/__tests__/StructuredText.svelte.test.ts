import { describe, expect, it } from 'vitest';
import '@testing-library/jest-dom';
import { render } from '@testing-library/svelte';

import {
	isBlock,
	isInlineBlock,
	isHeading,
	isInlineItem,
	isItemLink,
	isSpan
} from 'datocms-structured-text-utils';

import { StructuredText } from '../../..';

import {
	heading,
	paragraphWithLink,
	structuredTextWithBlocksAndLinks,
	full
} from './__fixtures__/structuredText';

import CustomSpan from './__fixtures__/CustomSpan.svelte';
import IncreasedLevelHeading from './__fixtures__/IncreasedLevelHeading.svelte';
import ItemLink from './__fixtures__/ItemLink.svelte';
import Block from './__fixtures__/Block.svelte';
import InlineItem from './__fixtures__/InlineItem.svelte';
import InlineBlock from './__fixtures__/InlineBlock.svelte';

describe('StructuredText', () => {
	describe('with no value', () => {
		it('renders null', () => {
			const { container } = render(StructuredText, { props: { data: null } });

			expect(container).toMatchSnapshot();
		});
	});

	describe('with a very simple pure dast (only the `value` of a structured text)', () => {
		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, { props: { data: heading.value } });

				expect(container).toMatchSnapshot();
			});
		});

		describe('with custom mark rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: { data: heading, components: [[isSpan, CustomSpan]] }
				});

				expect(container).toMatchSnapshot();
			});
		});
	});

	describe('with a very simple dast', () => {
		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, { props: { data: heading } });

				expect(container).toMatchSnapshot();
			});
		});

		describe('with custom mark rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: { data: heading, components: [[isSpan, CustomSpan]] }
				});

				expect(container).toMatchSnapshot();
			});
		});
	});

	describe('with a dast which has a link inside', () => {
		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, { props: { data: paragraphWithLink } });

				expect(container).toMatchSnapshot();
			});
		});
	});

	describe('with a dast with no links nor blocks', () => {
		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: { data: heading }
				});

				expect(container).toMatchSnapshot();
			});
		});

		describe('with custom rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: { data: heading, components: [[isHeading, IncreasedLevelHeading]] }
				});

				expect(container).toMatchSnapshot();
			});
		});
	});

	describe('with a dast including links and blocks', () => {
		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: {
						data: structuredTextWithBlocksAndLinks,
						components: [
							[isInlineItem, InlineItem],
							[isItemLink, ItemLink],
							[isBlock, Block],
							[isInlineBlock, InlineBlock],
						]
					}
				});

				expect(container).toMatchSnapshot();
			});
		});

		describe('with missing component for item links', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: {
							data: structuredTextWithBlocksAndLinks,
							components: [
								[isBlock, Block],
								[isInlineBlock, InlineBlock],
								[isInlineItem, InlineItem]
							]
						}
					});
				}).toThrowErrorMatchingSnapshot();
			});
		});

		describe('with missing component for inline items', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: {
							data: structuredTextWithBlocksAndLinks,
							components: [
								[isBlock, Block],
								[isInlineBlock, InlineBlock],
								[isItemLink, ItemLink]
							]
						}
					});
				}).toThrowErrorMatchingSnapshot();
			});
		});

		describe('with missing links', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: {
							data: { ...structuredTextWithBlocksAndLinks, links: [] },
							components: [
								[isBlock, Block],
								[isInlineBlock, InlineBlock],
								[isInlineItem, InlineItem],
								[isItemLink, ItemLink]
							]
						}
					});
				}).toThrowErrorMatchingSnapshot();
			});
		});

		describe('with missing component for blocks', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: {
							data: structuredTextWithBlocksAndLinks,
							components: [
								[isInlineItem, InlineItem],
								[isItemLink, ItemLink]
							]
						}
					});
				}).toThrowErrorMatchingSnapshot();
			});
		});

		describe('with missing component for inline blocks', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: {
							data: structuredTextWithBlocksAndLinks,
							components: [
								[isBlock, Block],
								[isInlineItem, InlineItem],
								[isItemLink, ItemLink]
							]
						}
					});
				}).toThrowErrorMatchingSnapshot();
			});
		});

		describe('with missing blocks', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: {
							data: { ...structuredTextWithBlocksAndLinks, blocks: [] },
							components: [
								[isBlock, Block],
								[isInlineBlock, InlineBlock],
								[isInlineItem, InlineItem],
								[isItemLink, ItemLink]
							]
						}
					});
				}).toThrowErrorMatchingSnapshot();
			});
		});
	});

	describe('with a dast including a sprinkle of everything', () => {
		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: {
						data: full,
						components: [
							[isInlineItem, InlineItem],
							[isItemLink, ItemLink],
							[isBlock, Block]
						]
					}
				});

				expect(container).toMatchSnapshot();
			});
		});
	});
});
