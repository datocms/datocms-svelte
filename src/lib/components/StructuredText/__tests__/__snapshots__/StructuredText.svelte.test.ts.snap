// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StructuredText > with a dast including a sprinkle of everything > with default rules > renders the document 1`] = `
<body>
  <div />
  <div />
  <div />
  <div />
  <div />
  <div />
  <div>
    <h1>
      This is a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <mark>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </mark>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      .
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      And it's berry perfect.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h2>
      Seriously, what the grape else do you want?
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      You probably build websites and think your banana is special.
      <br />
      You think your 13 megabyte parallax-ative home page is going to get you
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
         some fruity Awwward banner y
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      ou can glue to the top corner of your site. You think your 40-pound jQuery file and 83 polyfills give IE7 a peach because it finally has box-shadow. Wrong, melonhead. Let me describe your perfect-pear website:
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <ul>
      <li>
        <p>
          Fruit's lightweight and loads fast
          
          <!--&lt;Lines&gt;-->
          
          <!--&lt;Span&gt;-->
          
          
          
          <!--&lt;Node&gt;-->
          
          
        </p>
        <!--&lt;Paragraph&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        <ul>
          <li>
            <p>
              Looks the same in all your citrusy browsers
              
              <!--&lt;Lines&gt;-->
              
              <!--&lt;Span&gt;-->
              
              
              
              <!--&lt;Node&gt;-->
              
              
            </p>
            <!--&lt;Paragraph&gt;-->
            
            
            
            <!--&lt;Node&gt;-->
            
            
          </li>
          <!--&lt;ListItem&gt;-->
          
          
          
          <!--&lt;Node&gt;-->
          <li>
            <p>
              Fits on all your juicy screens
              
              <!--&lt;Lines&gt;-->
              
              <!--&lt;Span&gt;-->
              
              
              
              <!--&lt;Node&gt;-->
              
              
            </p>
            <!--&lt;Paragraph&gt;-->
            
            
            
            <!--&lt;Node&gt;-->
            <ul>
              <li>
                <p>
                  The kumquat's accessible to every apricot that visits your site
                  
                  <!--&lt;Lines&gt;-->
                  
                  <!--&lt;Span&gt;-->
                  
                  
                  
                  <!--&lt;Node&gt;-->
                  
                  
                </p>
                <!--&lt;Paragraph&gt;-->
                
                
                
                <!--&lt;Node&gt;-->
                
                
              </li>
              <!--&lt;ListItem&gt;-->
              
              
              
              <!--&lt;Node&gt;-->
              
              
            </ul>
            
            <!--&lt;List&gt;-->
            
            
            
            <!--&lt;Node&gt;-->
            
            
          </li>
          <!--&lt;ListItem&gt;-->
          
          
          
          <!--&lt;Node&gt;-->
          
          
        </ul>
        
        <!--&lt;List&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </li>
      <!--&lt;ListItem&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <li>
        <p>
          Fruit's legible and gets your juicy point across (if you had one instead of just 5mb pics of hipsters drinking coffee)
          
          <!--&lt;Lines&gt;-->
          
          <!--&lt;Span&gt;-->
          
          
          
          <!--&lt;Node&gt;-->
          
          
        </p>
        <!--&lt;Paragraph&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </li>
      <!--&lt;ListItem&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </ul>
    
    <!--&lt;List&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h3>
      Well guess what, you absolute kiwi:
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h3>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      You. Are. Over-designing. Look at this fruit. It's a banana-loving website. Why the fig do you need to animate a fruity trendy-pear banner flag when I hover over that useless piece of mango? You spent hours on it and added 80 kilobytes to your juicy site, and some papaya jabbing at it on their iPad with fat sausage fingers will never see that fruit. Not to mention blind people will never see that fruit, but they don't see any of your juicy fruit.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      You never knew it, but this is your perfect website. Here's why.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h2>
      It's fruity lightweight
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      This entire page weighs less than the gradient-meshed facebook logo on your juicy Wordpress site. Did you seriously load 100kb of jQuery UI just so you could animate the fruity background color of a div? You loaded all 7 fontfaces of a citrusy webfont just so you could say "Hi." at 100px height at the beginning of your site? 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      You piece of fruit.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <ul>
      <li>
        <p>
          
          <!--&lt;InlineItem&gt;-->
          
          
          
          <!--&lt;Node&gt;-->
          
          <!--&lt;ItemLink&gt;-->
          
          
          
          <!--&lt;Node&gt;-->
          
          
        </p>
        <!--&lt;Paragraph&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </li>
      <!--&lt;ListItem&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <li>
        <p>
          
          
          <!--&lt;Lines&gt;-->
          
          <!--&lt;Span&gt;-->
          
          
          
          <!--&lt;Node&gt;-->
          
          
        </p>
        <!--&lt;Paragraph&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </li>
      <!--&lt;ListItem&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </ul>
    
    <!--&lt;List&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;Block&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h2>
      It's responsive
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      You kumquat. You thought you needed media queries to be responsive, but no. Responsive means that it responds to whatever fruity screensize it's viewed on. This site doesn't care if you're on an iMac or a banana-loving Tamagotchi.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h2>
      It fruity works
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      Look at this fruit. You can read it ... that is, if you can read, melonhead. It makes sense. It has banana-loving hierarchy. It's using HTML5 tags so you and your pear-headed browser know what the fig's in this juicy site. That's semantics, you absolute kiwi.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      It has content on the fruity screen. Your site has three bylines and link to your dribbble account, but you spread it over 7 full screens and make me click some bobbing button to show me how cool the jQuery ScrollTo plugin is.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      Cross-browser compatibility? Load this kumquat in IE6. I fruity dare you.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h2>
      This is a website. 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <mark>
        Look at it.
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </mark>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
       You've never seen one before.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      Like the man who's never grown out his beard has no idea what his true natural state is, you have no fruity idea what a website is. All you have ever seen are citrusy skeuomorphic bastardizations of what should be text communicating a juicy message. This is a real, naked website. Look at it. It's fruity beautiful.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h3>
      Yes, this is fruity satire, you fig
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h3>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      I'm not actually saying your citrusy site should look like this. What I'm saying is that all the problems we have with websites are ones we create ourselves. Websites aren't broken by default, they are functional, high-performing, and accessible. You break them. You son-of-a-peach.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <blockquote>
      <p>
        "Good design is as little design as possible." 
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </p>
      <!--&lt;Paragraph&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
       
      <footer>
        some German fruithead
      </footer>
    </blockquote>
    <!--&lt;Blockquote&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <h2>
      Epilogue
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <p>
      From the philosophies expressed (poorly) above, 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="http://txti.es"
        rel="nofollow"
        target="_blank"
      >
        txti
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </a>
      <!--&lt;Link&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
       was created. You should try it today to make your own banana-loving websites.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a dast including links and blocks > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This is a
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <mark>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </mark>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="/docs/how-to-code"
      >
        How to code
      </a>
      ;
      
      <!--&lt;InlineItem&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="/docs/how-to-code"
        target="_blank"
      >
        here!
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </a>
      
      <!--&lt;ItemLink&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <em>
        Jane Doe
      </em>
      
      <!--&lt;InlineBlock&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <figure>
      <blockquote>
        Foo bar.
      </blockquote>
       
      <figcaption>
        Mark Smith
      </figcaption>
    </figure>
    
    <!--&lt;Block&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a dast including links and blocks > with missing blocks > raises an error 1`] = `[Error: The Structured Text document contains a 'inlineBlock' node, but cannot find a record with ID 789 inside data.blocks!]`;

exports[`StructuredText > with a dast including links and blocks > with missing component for blocks > raises an error 1`] = `[Error: The Structured Text document contains an 'inlineBlock' node, but no component for rendering is specified!]`;

exports[`StructuredText > with a dast including links and blocks > with missing component for inline blocks > raises an error 1`] = `[Error: The Structured Text document contains an 'inlineBlock' node, but no component for rendering is specified!]`;

exports[`StructuredText > with a dast including links and blocks > with missing component for inline items > raises an error 1`] = `[Error: The Structured Text document contains an 'inlineItem' node, but no component for rendering is specified!]`;

exports[`StructuredText > with a dast including links and blocks > with missing component for item links > raises an error 1`] = `[Error: The Structured Text document contains an 'itemLink' node, but no component for rendering is specified!]`;

exports[`StructuredText > with a dast including links and blocks > with missing links > raises an error 1`] = `[Error: The Structured Text document contains an 'itemLink' node, but cannot find a record with ID 123 inside data.links!]`;

exports[`StructuredText > with a dast which has a link inside > with default rules > renders the document 1`] = `
<body>
  <div>
    <p>
      This is a paragraph with a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="https://random.link"
        rel="nofollow"
        target="_blank"
      >
        berry cool link
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </a>
      <!--&lt;Link&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      . And the link has the target and the rel attributes set.
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </p>
    <!--&lt;Paragraph&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a dast with no links nor blocks > with custom rules > renders the document 1`] = `
<body>
  <div>
    <h2>
      This
      <br />
      is a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;IncreasedLevelHeading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a dast with no links nor blocks > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
      <br />
      is a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple dast > with custom mark rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
      <br />
      is a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <b>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;CustomSpan&gt;-->
      </b>
      
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple dast > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
      <br />
      is a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple pure dast (only the \`value\` of a structured text) > with custom mark rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
      <br />
      is a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <b>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;CustomSpan&gt;-->
      </b>
      
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple pure dast (only the \`value\` of a structured text) > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
      <br />
      is a 
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with no value > renders null 1`] = `
<body>
  <div>
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;
